pkg/metric_generator/generator.go

```diff
@@ -101,3 +101,26 @@ func FilterMetricFamilies(l allowDenyLister, families []FamilyGenerator) []Famil
 
 	return filtered
 }
+
+// FilterMetricFamiliesLabels takes a AllowLabels and a slice of metric
+// families and returns a filtered slice with metric families with their functions adapted to include only allowed labels for that metric Family.
+func FilterMetricFamiliesLabels(allowLabels AllowLabels, families []FamilyGenerator) []FamilyGenerator {
+	var filtered []FamilyGenerator
+	for _, f := range families {
+		generateFunc := func(familyGenerator FamilyGenerator) func(obj interface{}) *metric.Family {
+			return func(obj interface{}) *metric.Family {
+				metricFamily := familyGenerator.GenerateFunc(obj)
```

Comment:
With the current implementation if we call GenerateFunc on an object having the following label: `app.kubernetes.io/name` it will be converted to the Prometheus format: `label_app_kubernetes_io_name`. This means that if we have the following option passed to ksm: `--labels-allow-list="example_labels=[app.kubernetes.io/name]"`, the label will not be allowed.

Your approach is correct; however, if we want to be able to handle Kubernetes annotations and labels some rework needs to be done around the GenerateFunc and more precisely the `kubeLabelsToPrometheusLabels` calls.

Reply:
WDYT about a quick win and just pass the labels allow list to `kubeLabelsToPrometheusLabels` function?

Reply:
> WDYT about a quick win and just pass the labels allow list to kubeLabelsToPrometheusLabels function?

I think this makes sense, it will be the behavior users are most likely to expect.

Reply:
The problem with this approach is that we prefix labels coming from label related metrics.

I have a few open questions/thoughts...

Should we check if the metric is of a certain kind and add the prefix in the filtering stage/allow list stage rather than in the resource generate functions?

Or do we expect users to provide this type of value to the allowList flag ex: `label.labelname`  which technically would be just `labelname` label in k8s terms.

Reply:
@brancz